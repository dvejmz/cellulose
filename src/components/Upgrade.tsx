import React from 'react';

export interface UpgradeProps {
  name: string;
  cost: number;
  currency: string;
};

const Upgrade: React.FC<UpgradeProps> = (props: UpgradeProps) => (
  <div className="upgrade" data-test-id="upgrade-ppc-2x">
    <button
      className="btn"
    >
      {props.name} ({props.currency}{props.cost})
    </button>
  </div>
);

export default Upgrade;
