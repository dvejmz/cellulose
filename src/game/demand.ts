export const getDemand = (
  highDemandPrice: number,
  highDemandQty: number,
  lowDemandPrice: number,
  lowDemandQty: number,
  currentPrice: number
): number => {
  const higherDemandParams = { price: highDemandPrice, demand: highDemandQty };
  const lowerDemandParams = { price: lowDemandPrice, demand: lowDemandQty };

  const slope = (higherDemandParams.demand - lowerDemandParams.demand) / (higherDemandParams.price - lowerDemandParams.price);
  const offset = -(slope * higherDemandParams.price) + higherDemandParams.demand;
  return Math.max(slope * currentPrice + offset, 0);
};

export const getPurchaseRateFromDemand = (prevDemand: number, newDemand: number, prevPrice: number, newPrice: number): number => {
  const changeInQtyPct = ((newDemand - prevDemand) / ((newDemand + prevDemand) / 2)) * 100;
  const changeInPricePct = ((newPrice - prevPrice) / ((newPrice + prevPrice) / 2)) * 100;
  const priceElasticityDemand = changeInQtyPct / changeInPricePct;
  return isNaN(priceElasticityDemand) || !isFinite(priceElasticityDemand) ? 0 : 1 / Math.abs(priceElasticityDemand);
};
