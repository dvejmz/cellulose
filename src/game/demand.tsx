export const getDemand = (
  buyFactor: number,
  demandSlope: number,
  price: number
): number => (
  buyFactor - (demandSlope * price)
);

export const getPurchaseRateFromDemand = (demand: number): number => (
  demand / 10.0
);
