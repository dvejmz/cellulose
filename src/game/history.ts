interface PaperHistory {
  purchaseRate: number[];
};

export interface History {
  paper: PaperHistory;
  funds: number[];
};


