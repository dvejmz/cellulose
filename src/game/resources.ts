export interface Resources {
  paper: Resource;
  pulp: Resource;
}

export interface Resource {
  name: string;
  quantity: number;
  quantityUnit: string;
  price: number;
  purchaseRate: number;
}
