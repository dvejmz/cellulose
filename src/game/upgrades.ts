export interface Upgrade {
  id: string;
  previousId: string | null;
  name: string;
  cost: number;
  unlockCost: number;
  enabled: boolean;
};

export interface Upgrades {
  totalPaper: number;
  upgrades: Upgrade[];
};

export const UPGRADE_UNLOCK_TIER_1 = 100;

export const UPGRADE_ID_PPC_2X = 'upgrade-ppc-2x';
export const UPGRADE_ID_PPC_4X = 'upgrade-ppc-4x';
export const UPGRADE_ID_PPC_8X = 'upgrade-ppc-8x';
export const UPGRADE_ID_PPC_16X = 'upgrade-ppc-16x';

export const UPGRADE_ID_PAPERMAKER_1X = 'upgrade-papermaker-1x';

export const getUpgradeById = (id: string, upgrades: Upgrade[]): (Upgrade | undefined) => (
  [ ...upgrades ].find(u => u.id === id)
);

export const getInactiveUpgrades = (upgrades: Upgrade[]): Upgrade[] => (
  [ ...upgrades ].filter(u => !u.enabled)
);

const getActiveUpgrades = (upgrades: Upgrade[]) =>
  upgrades.filter(u => u.enabled).sort((lhs, rhs) => rhs.cost - lhs.cost);

// just fetch a list of everything under < UC which is not enabled and then make a reverse linked list with references
// to the previous, active upgrade for a possibly unlockable upgrade
export const getUnlockableUpgrades = (paperQty: number, upgrades: Upgrade[]): Upgrade[]  => {
  const activeUpgrades = getActiveUpgrades(upgrades);
  const inactiveUpgrades = getInactiveUpgrades(upgrades).filter(u => u.unlockCost <= paperQty);
  return inactiveUpgrades
    .filter(
      iu => (iu.previousId === null
              || activeUpgrades.findIndex(au => au.id === iu.previousId)) !== -1);
};

export const getActivePpcMultiplier = (upgrades: Upgrade[]): number => {
  if (!upgrades.length) {
    return 1;
  }

  const activeUpgrades = getActiveUpgrades(upgrades);
  if (!activeUpgrades.length) {
    return 1;
  }

  const highestActiveUpgrade = activeUpgrades[0];
  switch (highestActiveUpgrade.id) {
    case UPGRADE_ID_PPC_2X:
      return 2;
    case UPGRADE_ID_PPC_4X:
      return 4;
    case UPGRADE_ID_PPC_8X:
      return 8;
    case UPGRADE_ID_PPC_16X:
      return 16;
    default:
      return 1;
  }
};
