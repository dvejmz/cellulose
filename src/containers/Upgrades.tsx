import React from 'react';
import Upgrade from '../components/Upgrade';
import { Upgrade as UpgradeType } from '../game/upgrades';
import { getUnlockableUpgrades } from '../game/upgrades';

interface UpgradesProps {
  upgrades: UpgradeType[];
  totalPaper: number;
  currency: string;
  onBuyClick: (id: string, cost: number) => void;
};

const Upgrades: React.FC<UpgradesProps> = (props: UpgradesProps) => {
  const { upgrades, totalPaper, currency } = props;
  return (
    <div className="upgrades">
      {getUnlockableUpgrades(totalPaper, upgrades)
        .map(u => (
          <Upgrade
            id={u.id}
            key={u.id}
            name={u.name}
            cost={u.cost}
            currency={currency}
            onBuyClick={() => props.onBuyClick(u.id, u.cost)}
          />
        ))}
    </div>
  );
};

export default Upgrades;
