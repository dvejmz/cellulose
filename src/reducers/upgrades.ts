import * as Actions from '../actions';
import { RootReducerAction, Upgrades } from '../App';
import { getUpgradeById } from '../game/upgrades';

const upgradesReducer = (currentUpgrades: Upgrades, action: RootReducerAction): Upgrades => {
  switch (action.type) {
    case Actions.RESOURCES_MAKE_PAPER:
      return {
        ...currentUpgrades,
        totalPaper: currentUpgrades.totalPaper + 1,
      };
    case Actions.UPGRADES_BUY:
      const updatedUpgrades = [...currentUpgrades.upgrades];
      const upgradeIndex = updatedUpgrades.findIndex(u => u.id === action.data.id);
      if (upgradeIndex !== -1) {
        updatedUpgrades[upgradeIndex].enabled = true;
      }

      return {
        ...currentUpgrades,
        upgrades: updatedUpgrades,
      };
    default:
      return currentUpgrades;
  }
};

export default upgradesReducer;
