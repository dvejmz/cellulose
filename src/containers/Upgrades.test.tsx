import { mount, ReactWrapper } from 'enzyme';
import React from 'react';
import Upgrade from '../components/Upgrade';
import { UpgradeType } from '../game/upgrades';
import { getMockUpgrade } from '../utils/test/mocks';
import Upgrades from './Upgrades';

describe('<Upgrades />', () => {
  let wrapper: ReactWrapper;
  let upgrades: UpgradeType[];

  beforeEach(() => {
    upgrades = [
      getMockUpgrade({ id: 'upgrade-ppc-2x', unlockCost: 10 }),
      getMockUpgrade({ id: 'upgrade-ppc-4x', name: '4x PPC', unlockCost: 50 }),
      getMockUpgrade({ id: 'upgrade-ppc-4x', name: '8x PPC', unlockCost: 50, enabled: true }),
      getMockUpgrade({ id: 'upgrade-ppc-16x', name: '16x PPC', unlockCost: 110 }),
    ];
    wrapper = mount(
      <Upgrades
      upgrades={upgrades}
      totalPaper={100}
      currency="Â£"
      onBuyClick={() => {}} // tslint:disable-line no-empty
      />);
  });
  
  it('renders', () => {
    expect(upgrades).toMatchSnapshot();
  });

  it('only shows inactive upgrades with a paper cost less or equal than current paper', () => {
    const actualUpgrades = wrapper.find(Upgrade);
    expect(actualUpgrades).toHaveLength(2);
    expect(actualUpgrades.at(0).text()).toContain('2x PPC');
    expect(actualUpgrades.at(1).text()).toContain('4x PPC');
  });
});
