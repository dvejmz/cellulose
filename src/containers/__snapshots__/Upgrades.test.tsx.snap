// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Upgrades /> renders 1`] = `
Array [
  Object {
    "cost": 100,
    "enabled": false,
    "id": "upgrade-ppc-2x",
    "name": "2x PPC",
    "unlockCost": 10,
  },
  Object {
    "cost": 100,
    "enabled": false,
    "id": "upgrade-ppc-4x",
    "name": "4x PPC",
    "unlockCost": 50,
  },
  Object {
    "cost": 100,
    "enabled": true,
    "id": "upgrade-ppc-4x",
    "name": "8x PPC",
    "unlockCost": 50,
  },
  Object {
    "cost": 100,
    "enabled": false,
    "id": "upgrade-ppc-16x",
    "name": "16x PPC",
    "unlockCost": 110,
  },
]
`;
